### PR TITLE
`Development`: Document playwright setup and test parallelism

### PR DESCRIPTION
**Pull request by muradium**
#### General
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

### Motivation and Context
Developers need documentation about how to setup and run Playwright tests.

### Description
- Adds instructions for setting up Playwright and running e2e tests
- Explains how test parallelization works in Playwright and how to configure it

### Steps for Testing
Visit the changed documentation:
- [E2E Testing with Playwright](https://artemis-platform--8672.org.readthedocs.build/en/8672/dev/playwright.html)

#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### CODE REVIEW

**Review by sarpsahinalp — DISMISSED** [review]
lgtm but I might be biased since I worked with playwright tests before

File: docs/dev/playwright.rst [hunk 1]
```diff
@@ -0,0 +1,75 @@
+E2E Testing with Playwright
+===========================
+
+Set up Playwright locally
+-------------------------
+
+To run the tests locally, developers need to set up Playwright on their machine.
+End-to-end tests test entire workflows, therefore they require an entire Artemis setup - database, client and server
+to be running.
```
Comment (laadvo) [inline docs/dev/playwright.rst:9]:
```suggestion
End-to-end tests test entire workflows; therefore, they require the whole Artemis setup - database, client, and server to be running. 
```

File: docs/dev/playwright.rst [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+Playwright tests rely on the Playwright Node.js library, Playwright browser extensions and some helper packages.
```
Comment (laadvo) [inline docs/dev/playwright.rst:10]:
```suggestion
Playwright tests rely on the Playwright Node.js library, Playwright browser extensions, and some helper packages.
```

File: docs/dev/playwright.rst [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        BASE_URL=http://localhost:9000
+        EXERCISE_REPO_DIRECTORY=test-exercise-repos
+
+    Make sure ``BASE_URL`` matches your Artemis client URL and ``ADMIN_USERNAME`` and
+    ``ADMIN_PASSWORD`` match your Artemis admin user credentials.
+
+3. Configure test users
+
+     Playwright tests require users with different roles to simulate concurrent user interactions. You can configure
+     user ids and check their corresponding user roles in ``src/test/playwright/support/users.ts`` file. Usernames are
```
Comment (laadvo) [inline docs/dev/playwright.rst:42]:
```suggestion
     user IDs and check their corresponding user roles in the ``src/test/playwright/support/users.ts`` file. Usernames are
```

File: docs/dev/playwright.rst [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+     defined automatically by replacing the ``USERID`` part in ` ``PLAYWRIGHT_USERNAME_TEMPLATE`` with corresponding
+     user id. If users with such usernames do not exist, set ``CREATE_USERS`` to ``true`` on the ``playwright.env``
```
Comment (laadvo) [inline docs/dev/playwright.rst:44]:
```suggestion
     user ID. If users with such usernames do not exist, set ``CREATE_USERS`` to ``true`` on the ``playwright.env``
```

File: docs/dev/playwright.rst [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+     file for users to be created on setup stage. If there exist users with the same name(s) but different user role(s),
+     change user ids to different values to ensure new users are created with roles as defined in configuration.
```
Comment (laadvo) [inline docs/dev/playwright.rst:46]:
```suggestion
     file for users to be created on setup stage. If users with the same name(s) but different user role(s) exist,
     change user IDs to different values ensuring new users are created with roles as defined in configuration.
```

File: docs/dev/playwright.rst [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+4. Setup Playwright package and its browser extensions:
+
+    Run ``npm run playwright:setup`` to install Playwright browser binaries, set up the environment to ensure Playwright
+    can locate these binaries and create test users if creating users is enabled in configuration.
```
Comment (laadvo) [inline docs/dev/playwright.rst:51]:
```suggestion
    can locate these binaries and create test users (if creating users is enabled in configuration).
```

File: docs/dev/playwright.rst [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+5. Open Playwright UI
+
+    Run ``npm run playwright:open`` to open the Playwright UI. This is a graphical interface that allows you to run
+    individual tests, test files or test suites.
```
Comment (laadvo) [inline docs/dev/playwright.rst:56]:
```suggestion
   Run ``npm run playwright:open`` to open the Playwright UI. This graphical interface allows you to run individual tests, test files, or test suites.
```

File: docs/dev/playwright.rst [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    Another way to run tests is to use the command ``npm run playwright:run``. This command runs all tests in command
+    line.
+
+Test parallelization
+--------------------
+
+Running tests in parallel may speed up test execution. Playwright tests are configured to run in fully parallel mode
+by default. It means all tests in all files are executed in parallel. This is achieved by Playwright's built-in
+parallelization feature. Test execution tasks are divided among worker processes. Each process runs a separate browser
+instance and executes a subset of tests. The number of worker processes can be adjusted in ``playwright.config.js``
```
Comment (laadvo) [inline docs/dev/playwright.rst:66]:
```suggestion
instance and executes a subset of tests. The number of worker processes can be adjusted in the ``playwright.config.js``
```

File: docs/dev/playwright.rst [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+file.
+
+.. warning ::
+    Using more worker processes divides the available computing resources, giving each worker fewer resources. If you
+    use too many workers, it can lead to resource contention, slowing down individual test execution and potentially
+    causing timeouts.
+
+To make tests run sequentially, set the ``workers`` option to ``1``. If you want tests in a single file to run
```
Comment (laadvo) [inline docs/dev/playwright.rst:74]:
```suggestion
To run tests sequentially, set the ``workers`` option to ``1``. If you want tests in a single file to run
```

**Review by laadvo — COMMENTED** [review]
The documentation is written clearly and easy to understand even for someone who did not work with Playwright before. I just have a few suggestions on how to improve the grammar.

**Review by laadvo — APPROVED** [review]
Thanks for implementing my suggestions! LGTM

**Review by sarpsahinalp — DISMISSED** [review]
Reapprove

File: docs/dev/playwright.rst [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    Make sure ``BASE_URL`` matches your Artemis client URL and ``ADMIN_USERNAME`` and
+    ``ADMIN_PASSWORD`` match your Artemis admin user credentials.
+
+3. Configure test users
+
+     Playwright tests require users with different roles to simulate concurrent user interactions. You can configure
+     user IDs and check their corresponding user roles in the ``src/test/playwright/support/users.ts`` file. Usernames are
+     defined automatically by replacing the ``USERID`` part in ` ``PLAYWRIGHT_USERNAME_TEMPLATE`` with corresponding
+     user ID. If users with such usernames do not exist, set ``CREATE_USERS`` to ``true`` on the ``playwright.env``
+     file for users to be created on setup stage. If users with the same name(s) but different user role(s) exist,
```
Comment (pzdr7) [inline docs/dev/playwright.rst:44]:
```suggestion
     file for users to be created during the setup stage. If users with the same usernames but different user roles already exist,
```

File: docs/dev/playwright.rst [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+     change user IDs to different values ensuring new users are created with roles as defined in configuration.
```
Comment (pzdr7) [inline docs/dev/playwright.rst:45]:
```suggestion
     change the user IDs to different values to ensure that new users are created with roles as defined in the configuration.
```

**Review by az108 — COMMENTED** [review]
LGTM aswell. Will approve after the feedback mentioned before is implemented :)

**Review by az108 — APPROVED** [review]
Reapprove

**Review by pzdr7 — APPROVED** [review]
Thanks for including the changes; LGTM 👍

**Review by laadvo — APPROVED** [review]
reapprove

**Review by sarpsahinalp — APPROVED** [review]
Reapprove

**Review by DominikRemo — APPROVED** [review]
Reapprove

**Pull request merged**